### PR TITLE
Merging dev/Api to Master

### PR DESCRIPTION
**Pull request by joperezr**
Merging dev/api branch to master.

The important part of this review is commit d81cdea472a16251ef6c78b7a6b751f30dd3424e which basically includes all of the changes I had to do in order for the dev/api code to build in master branch, and that includes:

- Rev'ing up the version of some contracts (System.Console, System.IO.FileSystem, System.Runtime.Extensions)
- Separating the new tests for the added API into a different test project that targets netstandard1.7. **FOR NOW THIS TESTS WILL NOT BE RUN ~~OR BUILT~~(<-EDIT:new tests will be built now), SINCE WE NEED A BUILDTOOLS UPDATE FOR THAT**
- updating dependency versions to match the expected ones.
- Fixing merge conflicts.

As you can see, the commit that was used as a base to merge the PR was 20b0dbde05e861225fa522e1e05cc52796932df8 which is outdated, the plan is to get this PR merged, so people can send out their new PRs against master, and have one last merge from whatever remains from dev/api to master again.

cc: @weshaggard @danmosemsft @ericstj @stephentoub 

FYI: @karajas

### CODE REVIEW

**Comment by danmosemsft** [comment]
What is the timeframe for the build tools update - can you make sure there's an issue?

LGTM for merging, if @ericstj signs off.

**Comment by stephentoub** [comment]
> Separating the new tests for the added API into a different test project that targets netstandard1.7

Why different test projects?  We can't just add test files to the same project, conditionally included?  I'd really like to avoid us having a multitude of test projects per assembly, split based purely on when certain APIs got added.

**Comment by joperezr** [comment]
@dotnet-bot test Innerloop CentOS7.1 Release Build and Test please

**Comment by joperezr** [comment]
@stephentoub I talked to @ericstj about this, and apparently it is not as straight forward as conditioning the `<Compile>` itemgroup depending on the TFM. There are more changes required in order to make sure that depending on the Compile-TFM (i.e. netstandard1.6) we run the tests on the correct Runtime-TFMs(i.e. netcoreapp1.0). Also, if we were to condition the `<Compile>` itemGroup we can't do something like `Condition="'$(TFM)'=='netstandard1.7'"` because when we ship netstandard1.8 and retarget then all of those conditions will be invalid. In any case, after that discussion we said that probably the best way to do this for now is to separate them to be able to get this merge in, and then log issues to work on the required infrastructure of trying to work this out differently.

Is that ok by you?

**Comment by stephentoub** [comment]
> Is that ok by you?

Thanks for the explanation.  As long as we follow-up to fix it subsequently, I'm ok with it.  I'd just really like this to not fall by the wayside.

**Comment by danmosemsft** [comment]
Just to be 100% explicit Jose, I'm reading this as (a) the existing tests will continue to run as normal (b) the new tests must be run manually by running msbuild in the folder. If so this is okay.

**Comment by joperezr** [comment]
@danmosemsft not exactly. It is more like:
- The existing tests will continue to run as normal
- New tests will get built, but not executed. If you try to execute them they won't be able to run, until we get a buildtools update that supports running tests for netstandard 1.7 AND we also have to do the work to get the implementation of the new libraries in the package. With this change I'm only adding the new contracts to the packages, which is why new tests can build fine, but won't be able to run.

**Comment by ericstj** [comment]
> running tests for netstandard 1.7 AND we also have to do the work to get the implementation of the new libraries in the package.

We don't test "netstandard" that's not a runnable thing.  We test on platforms that support netstandard.  In this case it'll be net463 and netcoreapp1.1.  I have a [branch](https://github.com/ericstj/buildtools/tree/updateNugetEnableHigherImpl) in buildtools that should enable building implementations that target these and validates ns1.7 api against those impls; I should have a PR ready today.  We still may need some more for testing /cc @karajas.

File: src/System.Json/pkg/System.Json.pkgproj [hunk 1]
```diff
@@ -0,0 +1,34 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <PropertyGroup>
+    <!-- Skipping SupportCheck for this package since we need to update the FrameworkList for this to work-->
+    <SkipSupportCheck>true</SkipSupportCheck>
```
Comment (ericstj) [inline src/System.Json/pkg/System.Json.pkgproj]: You should be able to remove this now and add an entry for $(AllXamarinFrameworks).  The buildtools update with this included went in yesterday.

File: src/System.Console/ref/System.Console.csproj [hunk 2]
```diff
@@ -2,9 +2,9 @@
 <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" DefaultTargets="Build">
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <PropertyGroup>
-    <AssemblyVersion>4.0.1.0</AssemblyVersion>
+    <AssemblyVersion>4.1.0.0</AssemblyVersion>
     <OutputType>Library</OutputType>
-    <PackageTargetFramework>netstandard1.3</PackageTargetFramework>
+    <PackageTargetFramework>netstandard1.7</PackageTargetFramework>
     <NuGetTargetMoniker>.NETStandard,Version=v1.3</NuGetTargetMoniker>
```
Comment (bartonjs) [inline src/System.Console/ref/System.Console.csproj]: @ericstj This seems off to me, can you confirm the validity?  (The NuGet moniker saying 1.3 but the TargetFramework saying 1.7)

Comment (ericstj) [inline src/System.Console/ref/System.Console.csproj]: Yes, these should be kept consistent.  Eventually I'd like to get rid of the duplication.

**Comment by weshaggard** [comment]
closing this PR as we aren't planning to do this merge just yet. We will do it post branching of master for the next release and once we get dev/api branch in a consistent state.

**Pull request closed without merging**